Add Mongo Connection Pool Metrics

Postremus opened the pull request:
Add metrics for opened connection pools in the mongo java driver.
Currently, the following metrics are available:
* size of a pool
* amount of currently used connections (checked-out-count)

See also: https://mongodb.github.io/mongo-java-driver/3.12/driver/reference/monitoring/

machi1990 commented on extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientProcessor.java:
```diff
@@ -24,20 +24,13 @@
-import io.quarkus.arc.deployment.GeneratedBeanBuildItem;
-import io.quarkus.arc.deployment.GeneratedBeanGizmoAdaptor;
-import io.quarkus.arc.deployment.UnremovableBeanBuildItem;
+import io.quarkus.arc.deployment.*;
```
can we explicit the imports in use and not use start imports here?

Postremus replied:
done

machi1990 commented on extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientProcessor.java:
```diff
@@ -24,20 +24,13 @@
-import io.quarkus.deployment.builditem.ExtensionSslNativeSupportBuildItem;
-import io.quarkus.deployment.builditem.FeatureBuildItem;
-import io.quarkus.deployment.builditem.SslNativeConfigBuildItem;
+import io.quarkus.deployment.builditem.*;
```
same here about the usage of star imports..

machi1990 commented on extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/metrics/MongoMetricsConnectionPoolListener.java:
```diff
@@ -0,0 +1,83 @@
+
+import com.mongodb.connection.ServerId;
+import com.mongodb.event.*;
+
```
let's get rid of wildcard imports..

machi1990 commented:
/cc @jmartisk this one might interest you :-)

jmartisk commented on docs/src/main/asciidoc/mongodb.adoc:
```diff
@@ -523,6 +523,14 @@ So when you access the `/health/ready` endpoint of your application you will hav
 
+== Metrics
+
+If you are using the `quarkus-smallrye-metrics` extension, `quarkus-mongodb` will automatically add metrics for the connection pools.
```
"Will automatically add metrics" followed by explanation that it must be explicitly enabled is a bit confusing.

jmartisk commented on extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/metrics/MongoMetricsConnectionPoolListener.java:
```diff
@@ -0,0 +1,95 @@
+        Tag[] tags = createTags(event.getConnectionId().getServerId());
+
+        Counter sizeCount = getMetricRegistry().counter(sizeName, tags);
+        sizeCount.inc(-1);
```
Counters are only supposed to go up. If you need a metric that goes down as well, it should be a `Gauge`. This would confuse Prometheus, which only expects counters to go up.

jmartisk commented on extensions/mongodb-client/deployment/src/test/java/io/quarkus/mongodb/MongoMetricsTest.java:
```diff
@@ -0,0 +1,65 @@
+        client = com.mongodb.client.MongoClients
+                .create(MongoClientSettings.builder().applyConnectionString(new ConnectionString(getConnectionString()))
+                        .applyToConnectionPoolSettings(
+                                builder -> builder.addConnectionPoolListener(new MongoMetricsConnectionPoolListener()))
```
IIUC this connection pool listener is needed for updating the metrics, but here you only explictly add it in a test, but how will it be added in production code? I suppose that when `quarkus.mongodb.metrics.enabled` is true, then we need to do this automatically. So something like this needs to be in the place in Quarkus codebase that produces Mongo clients (`AbstractMongoClientProducer` I think?) To be able to do that, `AbstractMongoClientProducer` will need to be able to see the value of `quarkus.mongodb.metrics.enabled` property, so I think it might need to be turned into a runtime property

jmartisk commented on extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/metrics/MongoMetricsConnectionPoolListener.java:
```diff
@@ -0,0 +1,95 @@
+        Metadata checkedOutCountMetadata = Metadata.builder().withName(checkedOutCountName).withType(MetricType.COUNTER)
+                .withDescription(" the current count of connections that are currently in use").build();
+
+        getMetricRegistry().register(sizeMetadata, new ConnectionPoolCounter(), tags);
```
In development mode, after a reload, the metrics from previous run may still exist in the registry. At the moment we don't automatically deregister vendor metrics during reload (I think we should, I will take care of that soon, but for now we don't). So temporarily this should probably contain a check if the metric already exists, and if yes, it probably means that we just reloaded and the metric should be removed and re-registered (so it will start again from zero).

jmartisk commented on extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientProcessor.java:
```diff
@@ -313,4 +315,13 @@ HealthBuildItem addHealthCheck(MongoClientBuildTimeConfig buildTimeConfig) {
+            MongoClientBuildTimeConfig buildTimeConfig, Capabilities capabilities) {
+
+        if (buildTimeConfig.metricsEnabled && capabilities.isCapabilityPresent(Capabilities.METRICS)) {
+            producer.produce(new SystemPropertyBuildItem("mongodb.metrics.enabled", "true"));
```
What is this for? Is it something that the Mongo client library recognizes? I haven't been able to find what it is used for at runtime.

jmartisk commented on extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientProcessor.java:
```diff
@@ -313,4 +315,13 @@ HealthBuildItem addHealthCheck(MongoClientBuildTimeConfig buildTimeConfig) {
+
+    @BuildStep
+    void produceMetricsEnabledProperty(BuildProducer<SystemPropertyBuildItem> producer,
+            MongoClientBuildTimeConfig buildTimeConfig, Capabilities capabilities) {
```
AFAIK the full list of metrics is known at build time? If so, we should register them here (including their tags). See https://github.com/quarkusio/quarkus/blob/master/extensions/agroal/deployment/src/main/java/io/quarkus/agroal/deployment/AgroalProcessor.java#L427 for example

Postremus replied:
> AFAIK the full list of metrics is known at build time?
We do not necessarily know all tags.
We use the tags host and port here.

In case of the mongo+srv protocoll, the servers are resolved through dns entries.
https://docs.mongodb.com/manual/reference/connection-string/#dns-seedlist-connection-format
We therefore would not know them during build time.

jmartisk replied:
Ok, that's unfortunate. In that case, the metrics won't currently get automatically deregistered during reload in dev mode (until we include a commit for https://github.com/quarkusio/quarkus/issues/7018), therefore `MongoMetricsConnectionPoolListener` will at least temporarily need to take into account the fact that after a reload, the metrics from previous run can still exist, and instead of just registering them, it should delete and re-register them if they exist, so that they can start again from zero

loicmathieu commented on extensions/mongodb-client/deployment/pom.xml:
```diff
@@ -58,6 +58,11 @@
         </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-smallrye-metrics-deployment</artifactId>
```
This should be `quarkus-smallrye-metrics-spi`

Postremus replied:
quarkus-smallrye-metrics-spi does not have the MetricRegistries class.
If I use an `@Inject RegistryType(VENDOR) MetricRegistry` for the Metric Registry, arc can not find an implementation.

How can I just use quarkus-smallrye-metrics-spi in my tests?

jmartisk replied:
`quarkus-smallrye-metrics-spi` basically only contains the `MetricBuildItem` class which is used for registering metrics at build time. If you need your tests to use the metrics extension, importing `quarkus-smallrye-metrics-deployment` is fine IMO. 
Instead of using `MetricRegistries`, which is a class from SmallRye Metrics, I'd suggest staying with just pure MP Metrics API, that is indeed `@Inject RegistryType(VENDOR) MetricRegistry`. If you import `quarkus-smallrye-metrics-deployment` it should work I think

loicmathieu replied:
> If you need your tests to actually use the metrics extension, importing quarkus-smallrye-metrics-deployment is fine IMO.

I'm not agree, if your test need `quarkus-smallrye-metrics-deployment` you should add it with the `test` scope and also import the SPI. With this, someone that don't want metrics can still avoid the inclusion of all the dependencies at runtime.

jmartisk replied:
- it has the `test` scope here
- What's the point in importing the SPI when you're not using anything that it offers?

loicmathieu replied:
@jmartisk sorry, I didn't notice that it has the test scope.

The fact that it didn't use the `MetricBuildItem` is a little annoying as it will be better than all extension register metrics the same way.

jmartisk replied:
I agree it would be convenient, and we try to do that as much as possible, but it's not achievable always. Not always do extensions know the full list of metrics at build time.

loicmathieu commented on extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientBuildTimeConfig.java:
```diff
@@ -11,4 +11,10 @@
+    /**
+     * Whether or not metrics are published in case the smallrye-metrics extension is present.
+     */
+    @ConfigItem(name = "metrics.enabled", defaultValue = "false")
```
`false` is the default value for boolean so it could be omitted

loicmathieu reviewed (changes requested):
+1 for registering the listener automatically at build time

Postremus commented:
> +1 for registering the listener automatically at build time

I have no idea how to do this. The MongoClientSettings class, and the MongoClientSettings.Builder is not serializeable. I therefore can't provide an instance of the settings directly using a build item. Any idea how to do this instead?

For now, I just produce a MongoConnectionPoolListenerBuildItem, containing the listener.

Postremus commented on extensions/mongodb-client/deployment/src/test/java/io/quarkus/mongodb/MongoMetricsTest.java:
```diff
@@ -0,0 +1,53 @@
+        //String name = client.listDatabaseNames().first();
+
+        //assertEquals(1L, getCounterValueOrNull("mongodb.connection-pool.size", getTags()));
+        //assertEquals(0L, getCounterValueOrNull("mongodb.connection-pool.checked-out-count", getTags()));
```
Yeah, I just noticed that I deactivated this test.

jmartisk commented:
The new solution that you pushed looks nice, that's what I was envisioning! 
I suppose that's what @loicmathieu  meant too, because you can't really register the listener at build time, if the connections are created  at runtime. 
From my PoV, if we just finish `MongoMetricsTest`, then we should be good to go.

loicmathieu commented on extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientProcessor.java:
```diff
@@ -75,9 +78,14 @@ UnremovableBeanBuildItem markBeansAsUnremovable() {
             CodecProviderBuildItem codecProvider,
             BsonDiscriminatorBuildItem bsonDiscriminator,
-            MongodbConfig config) {
+            MongodbConfig config, List<MongoConnectionPoolListenerBuildItem> connectionPoolListenerProvider) {
```
Can you move the connectionPoolListenerProvider on a new line, this will be easier to read

loicmathieu commented on extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientProcessor.java:
```diff
@@ -332,4 +340,14 @@ HealthBuildItem addHealthCheck(MongoClientBuildTimeConfig buildTimeConfig) {
     }
+
+    @BuildStep
+    void produceMongoClientSettings(
```
We usually name a build step method by what it does. Here you setup metrics or you add the listener so maybe `setupMetrics` or `addConnectionPoolListener` are better names.

loicmathieu commented on extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/metrics/MongoMetricsConnectionPoolListener.java:
```diff
@@ -0,0 +1,116 @@
+import io.smallrye.metrics.MetricRegistries;
+
+public class MongoMetricsConnectionPoolListener implements ConnectionPoolListener {
+    private final static String sizeName = "mongodb.connection-pool.size";
```
constants are usally defined as `private static final` and are uppercase.
```java
private static final String SIZE_PROPERTY = "mongodb.connection-pool.size";
```

Same for `checkedOutCountName`.

loicmathieu reviewed (approved):
I added some small code changes not mandatory so I aprove the PR.

Postremus commented:
ci failure seems unrelated

Postremus commented:
@cescoffier would you please take a look? Would be kind of nice to have this in 1.3.0.

Postremus commented:
I rebased and resolved the conflicts.

Current Build failure seems unrelated. Resteasy classes are missing.
`
[ERROR] testAggregate  Time elapsed: 0 s  <<< ERROR!
java.lang.NoClassDefFoundError: org/jboss/resteasy/core/ResteasyContext
	at org.jboss.resteasy.context.ResteasyContextProvider.currentContext(ResteasyContextProvider.java:18)
	at io.smallrye.context.impl.CapturedContextState.<init>(CapturedContextState.java:21)
	at io.smallrye.context.SmallRyeContextManager.captureContext(SmallRyeContextManager.java:64)
	at io.smallrye.context.SmallRyeThreadContext.currentContextExecutor(SmallRyeThreadContext.java:208)`

geoand commented on docs/src/main/asciidoc/mongodb.adoc:
```diff
@@ -543,6 +543,14 @@ So when you access the `/health/ready` endpoint of your application you will hav
 
+== Metrics
+
+If you are using the `quarkus-smallrye-metrics` extension, `quarkus-mongodb` con provide metrics about the connection pools.
```
`con`?

geoand commented on extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientProcessor.java:
```diff
@@ -114,7 +124,7 @@ BsonDiscriminatorBuildItem collectBsonDiscriminators(CombinedIndexBuildItem inde
      * Create a producer bean managing the lifecycle of the MongoClient.
      * <p>
      * The generated class will look like
-     * 
```
Any reason this change was made or what it simply done by the formatter? Dittor for the similar changes below

Postremus replied:
I think this might just have been to formatter, I will check.

geoand commented on extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/AbstractMongoClientProducer.java:
```diff
@@ -295,6 +297,13 @@ private MongoClientSettings createMongoConfiguration(MongoClientConfig config) {
         }
+
+        if (!connectionPoolListeners.isEmpty()) {
+            settings.applyToConnectionPoolSettings(builder -> {
```
Please change this to not use lambdas. We try very hard to not use lambdas in runtime code because a small startup time and have a small memory footprint

geoand commented on extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/AbstractMongoClientProducer.java:
```diff
@@ -295,6 +297,13 @@ private MongoClientSettings createMongoConfiguration(MongoClientConfig config) {
+
+        if (!connectionPoolListeners.isEmpty()) {
+            settings.applyToConnectionPoolSettings(builder -> {
+                connectionPoolListeners.forEach(builder::addConnectionPoolListener);
```
Same for method references :)

geoand reviewed (commented):
I added some minor comments. After addressing them please rebase onto master in order to pick the latest CI changes

Postremus committed:
Add Mongo Connection Pool Metrics
Add metrics for every opened connection pool regarding size and currently used connections

machi1990 commented:
Merged, thanks @Postremus
